fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -495,20 +496,100 @@ void FixedLagSmoother::transactionCallback(
   }
 }
 
-void FixedLagSmoother::setDiagnostics(diagnostic_updater::DiagnosticStatusWrapper& status)
+/**
+ * @brief Make a diagnostic_msgs::DiagnosticStatus message filling in the level and message
+ *
+ * @param[in] level   The diagnostic status level
+ * @param[in] message The diagnostic status message
+ */
+diagnostic_msgs::DiagnosticStatus makeDiagnosticStatus(const int8_t level, const std::string& message)
 {
-  Optimizer::setDiagnostics(status);
+  diagnostic_msgs::DiagnosticStatus status;
+
+  status.level = level;
+  status.message = message;
 
-  if (status.level == diagnostic_msgs::DiagnosticStatus::OK)
+  return status;
+}
+
+/**
+ * @brief Helper function to generate the diagnostic status for each optimization termination type
+ *
+ * The termination type -> diagnostic status mapping is as follows:
+ *
+ * - CONVERGENCE, USER_SUCCESS -> OK
+ * - NO_CONVERGENCE            -> WARN
+ * - FAILURE, USER_FAILURE     -> ERROR (default)
+ *
+ * @param[in] termination_type The optimization termination type
+ * @return The diagnostic status with the level and message corresponding to the optimization termination type
+ */
+diagnostic_msgs::DiagnosticStatus terminationTypeToDiagnosticStatus(const ceres::TerminationType termination_type)
+{
+  switch (termination_type)
   {
-    status.message = "FixedLagSmoother " + status.message;
+    case ceres::TerminationType::CONVERGENCE:
+    case ceres::TerminationType::USER_SUCCESS:
+      return makeDiagnosticStatus(diagnostic_msgs::DiagnosticStatus::OK, "Optimization converged");
+    case ceres::TerminationType::NO_CONVERGENCE:
+      return makeDiagnosticStatus(diagnostic_msgs::DiagnosticStatus::WARN, "Optimization didn't converged");
+    default:
+      return makeDiagnosticStatus(diagnostic_msgs::DiagnosticStatus::ERROR, "Optimization failed");
   }
+}
+
+void FixedLagSmoother::setDiagnostics(diagnostic_updater::DiagnosticStatusWrapper& status)
+{
+  Optimizer::setDiagnostics(status);
 
-  status.add("Started", started_);
+  // Load std::atomic<bool> flag that indicates whether the optimizer has started or not
+  const bool started = started_;
+
+  status.add("Started", started);
   {
     std::lock_guard<std::mutex> lock(pending_transactions_mutex_);
     status.add("Pending Transactions", pending_transactions_.size());
   }
+
+  if (started)
+  {
+    // Add some optimization summary report fields to the diagnostics status if the optimizer has started
+    if (optimization_mutex_.try_lock())
```

Comment:
I use `try_lock()` here and below. I believe this is important. Otherwise, we could wait on the mutex, so the diagnostics contain all the fields every time.

Reply:
Fair enough. Done.

I made the `std::unique_lock<std::mutex> lock` const, to highlight we don't try unlock it manually.